src/keys/FileKey.cpp

```diff
@@ -218,87 +274,81 @@ bool FileKey::create(const QString& fileName, QString* errorMsg, int size)
  *
  * @param device input device
  * @return true on success
- * @deprecated
  */
-bool FileKey::loadXml(QIODevice* device)
+bool FileKey::loadXml(QIODevice* device, QString* errorMsg)
 {
     QXmlStreamReader xmlReader(device);
 
-    if (!xmlReader.error() && xmlReader.readNextStartElement()) {
-        if (xmlReader.name() != "KeyFile") {
-            return false;
-        }
-    } else {
+    if (xmlReader.error()) {
+        return false;
+    }
+    if (xmlReader.readNextStartElement() && xmlReader.name() != "KeyFile") {
         return false;
     }
 
-    bool correctMeta = false;
-    QByteArray data;
+    struct
+    {
+        QString version;
+        QByteArray hash;
+        QByteArray data;
+    } keyFileData;
 
     while (!xmlReader.error() && xmlReader.readNextStartElement()) {
         if (xmlReader.name() == "Meta") {
-            correctMeta = loadXmlMeta(xmlReader);
+            while (!xmlReader.error() && xmlReader.readNextStartElement()) {
+                if (xmlReader.name() == "Version") {
+                    keyFileData.version = xmlReader.readElementText();
+                    if (keyFileData.version == "1.00") {
+                        m_type = KeePass2XML;
+                    } else if (keyFileData.version == "2.0") {
+                        m_type = KeePass2XMLv2;
+                    } else {
+                        if (errorMsg) {
+                            *errorMsg = QObject::tr("Unsupported key file version: %1").arg(keyFileData.version);
+                        }
+                        return false;
+                    }
+                }
+            }
         } else if (xmlReader.name() == "Key") {
-            data = loadXmlKey(xmlReader);
+            while (!xmlReader.error() && xmlReader.readNextStartElement()) {
+                if (xmlReader.name() == "Data") {
+                    keyFileData.hash = QByteArray::fromHex(xmlReader.attributes().value("Hash").toLatin1());
+                    QByteArray rawData = xmlReader.readElementText().simplified().replace(" ", "").toLatin1();
+
+                    if (keyFileData.version == "1.00" && Tools::isBase64(rawData)) {
```

Comment:
We should add error message for the "else" case, neither the version matches or the raw data is not base64/hex.

Reply:
In that case it would fall through and test the next format.

Reply:
I decided to go for an error message. May fail in the rare case that somebody was using an invalid key file that looks like an XML key file, but isn't, but overall it's cleaner.